README.md
Lockpick
=
Lockpick is a ground-up C++17 rewrite of homebrew key derivation software, namely [kezplez-nx](https://github.com/tesnos/kezplez-nx). It also dumps titlekeys. This will dump all keys through `*_key_05` on firmwares below `6.2.0` and through `*_key_06` on `6.2.0`.

Due to key generation changes introduced in `7.0.0`, Lockpick is not able to dump keys ending in 07 at all. Furthermore, unfortunately the public method to dump `tsec_root_key` is only available on firmware `6.2.0` so `7.x` consoles can only dump through keys ending in 05.

What this software does differently
=
* Dumps `titlekeys` and SD seed
* Dumps all keys through `6.2.0`
* Uses the superfast `xxHash` instead of `sha256` when searching exefs for keys for a ~5x speed improvement
* Gets all possible keys from running process memory - this means no need to decrypt `Package2` at all, let alone decompress `KIP`s
* Gets bis keys and `header_key` without `tsec`, `sbk`, `master_key_00` or `aes` sources. Shoutout to exelix11 for using this method in [SwitchThemeInjector](https://github.com/exelix11/SwitchThemeInjector)! Homebrew devs should be doing this instead of requiring users to provide key files!

Usage
=
1. Use [Hekate v4.5+](https://github.com/CTCaer/hekate/releases) to dump TSEC and fuses:
    1. Push hekate payload bin using [TegraRCMSmash](https://github.com/rajkosto/TegraRcmSmash)/[TegraRCMGUI](https://github.com/eliboa/TegraRcmGUI)/modchip/injector
    2. Using the `VOL` and `Power` buttons to navigate, select `Console info...`
    3. Select `Print fuse info` (_not_ `kfuse info`)
    4. Press `Power` to save fuse info to SD card
    5. Select `Print TSEC keys`
    6. Press `Power` to save TSEC keys to SD card
2. Launch CFW of choice
3. Open `Homebrew Menu`
4. Run `Lockpick`
5. Use the resulting `/switch/prod.keys` file as needed and rename if required by any software you're using

You may instead use [biskeydump](https://github.com/rajkosto/biskeydump) and dump to SD to get all keys prior to the 6.2.0 generation - all keys up to those ending in 05. Lockpick will dump all keys up to that point regardless which firmware it's run on.

Notes
=
* To get keys ending in 06, you must have firmware `6.2.0` installed
* No one knows `package1_key_06`, it's derived and erased fully within the encrypted TSEC payload. While there's a way to extricate `tsec_root_key` due to the way it's used, this is unfortunately not true of the `package1` key
* If for some reason you dump TSEC keys on `6.2.0` and not fuses (`secure_boot_key`) you will still get everything except any of the `package1` or keyblob keys (without `secure_boot_key`, you can't decrypt keyblobs and that's where `package1` keys live)

Building
=
Release built with [libnx release v2.4.0](https://github.com/switchbrew/libnx).

Uses `freetype` which comes with `switch-portlibs` via `devkitPro pacman`:
```
pacman -S libnx switch-portlibs
```
then run:
```
make
```
to build.

Special Thanks
=
* tèsnos! For making [kezplez-nx](https://github.com/tesnos/kezplez-nx), being an all-around cool and helpful person and open to my contributions, not to mention patient with my *enthusiasm*. kezplez taught me an absolute TON about homebrew.
* SciresM for [hactool](https://github.com/SciresM/hactool), containing to my knowledge the first public key derivation *software*, and for `get_titlekeys.py`
* roblabla for the original keys [gist](https://gist.github.com/roblabla/d8358ab058bbe3b00614740dcba4f208) and for believing in our habilities
* The folks in the [ReSwitched](https://reswitched.team/) Discord server for answering my innumerable questions while researching this (and having such a useful chat backlog!)
* The memory reading code from jakibaki's [sys-netcheat](https://github.com/jakibaki/sys-netcheat) was super useful for getting keys out of running process memory
* The System Save dumping methodology from Adubbz' [Compelled Disclosure](https://github.com/Adubbz/Compelled-Disclosure)
* Shouts out to fellow key derivers: shadowninja108 for [HACGUI](https://github.com/shadowninja108/HACGUI), Thealexbarney for [Libhac](https://github.com/Thealexbarney/LibHac), and [rajkosto](https://github.com/rajkosto/) :eyes:
* [misson2000](https://github.com/misson20000) for help with `std::invoke` to get the function timer working
* [Simon](https://github.com/simontime) for the `eticket_rsa_kek` derivation method and for suggesting invoking `spl` for faster titlekey derivation
* [SciresM](https://github.com/SciresM) for the libnx aes library
* The constantly-improving docs on [Switchbrew wiki](https://switchbrew.org/wiki/) and [libnx](https://switchbrew.github.io/libnx/files.html)
* Literally the friends I made along the way! I came to the scene late and I've still managed to meet some wonderful people :) Thanks for all the help testing, making suggestions, and cheerleading!

Licenses
=
* `es` ipc code is from [Tinfoil](https://github.com/Adubbz/Tinfoil) licensed under [MIT](https://github.com/Adubbz/Tinfoil/blob/master/LICENSE)
* `FatFs R0.13c` is located [here](http://elm-chan.org/fsw/ff/00index_e.html) and is licensed under its own [BSD-style license](source/fatfs/LICENSE.txt)
* Simple `xxHash` implementation is from [stbrumme](https://github.com/stbrumme/xxhash) licensed under [MIT](https://github.com/stbrumme/xxhash/blob/master/LICENSE)
* Padlock icon is from [Icons8](https://icons8.com/) licensed under [Creative Commons Attribution-NoDerivs 3.0 Unported](https://creativecommons.org/licenses/by-nd/3.0/)

changelog.md
# Changelog
## Version 1.2.6
* Fix bis key generation on newer hardware

## Version 1.2.5
* Support Hekate v5 fuse dump format
* Make names consistent with libnx v2.2.0
* Adjust text alignment and coloring in Lockpick_RCM note

## Version 1.2.4
* Support new emunand FS memory layout
* No longer save header_key if empty

## Version 1.2.3
* Remove mbedtls dependency in favor of new libnx crypto library
* Remove libnx 1.6.0 support since crypto requires later commit
* Skip contradictory messaging if skipping keyfile save

## Version 1.2.2
* Do not overwrite existing keyfile that contains master_key_07
* Read eticket_rsa_kek from existing keyfile in case user is only running this for titlekeys
* Create /switch folder if needed

## Version 1.2.1
* Generate bis keys without master keys
* Update file size check to support Hekate v4.8 TSEC dump
* Fixed prod.keys alphabetization error
* Fixed build warning for ff.c
* Added in-app disclaimer about which keys can be dumped

## Version 1.2
* Update for libnx v2.0.0 compatibility and still runs when built with v1.6.0
  * The binary got even smaller!
* Accelerate finding FS keys
  * No longer find BIS sources as they're hardcoded (whoops)
  * Find all keys on first pass hashing FS instead of hashing the whole thing from the beginning repeatedly (__*whoops*__)

## Version 1.1.1
* No longer try to dump SD seed and ES keys on 1.0.0 as they're not available until 2.0.0

## Version 1.1
* Changed titlekey dump methodology
  * No longer crashes sysmodule, reboot no longer needed
  * Queries ES to verify ticket list is accurate
  * May take slightly longer than before on systems with hundreds of tickets
* Now dumps SD seed
* Reorganized and clarified UI text
  * Now indicates if no titles are installed to dump titlekeys from
* Swapped C++ stream functions for C I/O to reclaim some speed and binary size
* Tightened up dependencies

source/Common.cpp
/*
 * Copyright (c) 2018 shchmue
 *
 * This program is free software; you can redistribute it and/or modify it
 * under the terms and conditions of the GNU General Public License,
 * version 2, as published by the Free Software Foundation.
 *
 * This program is distributed in the hope it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
 * more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

/*
    Shared font code adapted from
    https://github.com/switchbrew/switch-examples/tree/master/graphics/shared_font
*/

#include "Common.hpp"
#include "Key.hpp"

#include <machine/endian.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include <filesystem>
#include <string>

#include <switch.h>

namespace Common {
    static u32 framebuf_width = 0;
    static Framebuffer fb;
    static u32 stride;
    static u32 *framebuf;
    // FreeType vars
    static FT_Library library;
    static FT_Face face;

    void draw_glyph(FT_Bitmap *bitmap, u32 x, u32 y, u32 color) {
        u32 framex, framey;
        u8* imageptr = bitmap->buffer;

        for (u32 tmpy = 0; tmpy < bitmap->rows; tmpy++) {
            for (u32 tmpx = 0; tmpx < bitmap->width; tmpx++) {
                framex = x + tmpx;
                framey = y + tmpy;

                // color calculation is extremely simplistic and only supports 0 or 0xff in each element for proper antialiasing
                framebuf[framey * framebuf_width + framex] = RGBA8_MAXALPHA(imageptr[tmpx], imageptr[tmpx], imageptr[tmpx]) & color;
            }

            imageptr += bitmap->pitch;
        }
    }

    void draw_line(u32 x, u32 y, u32 length, u32 color) {
        for (u32 tmpx = x; tmpx < x + length; tmpx++)
            framebuf[y * framebuf_width + tmpx] = color;
    }

    void draw_set_rect(u32 x, u32 y, u32 width, u32 height, u32 color) {
        for (u32 tmpy = y; tmpy < y + height; tmpy++)
            draw_line(x, tmpy, width, color);
    }

    void draw_text(u32 x, u32 y, u32 color, const char *str) {
        u32 tmpx = x;
        FT_UInt glyph_index;
        FT_GlyphSlot slot = face->glyph;

        for (size_t i = 0; i < strlen(str); i++) {
            if (str[i] == '\n') {
                tmpx = x;
                y += face->size->metrics.height / 64;
                continue;
            }

            glyph_index = FT_Get_Char_Index(face, (FT_ULong)str[i]);

            if (R_FAILED(FT_Load_Glyph(face, glyph_index, FT_LOAD_COLOR) ||
                FT_Render_Glyph(face->glyph, FT_RENDER_MODE_NORMAL) != 0))
            {
                return;
            }

            draw_glyph(&slot->bitmap, tmpx + slot->bitmap_left, y - slot->bitmap_top, color);

            tmpx += slot->advance.x >> 6;
            y += slot->advance.y >> 6;
        }
    }

    void draw_text_with_time(u32 x, u32 y, u32 color, const char *str, const float time) {
        char time_str[32];
        sprintf(time_str, "%1.6f seconds", time);
        draw_text(x, y, color, str);
        draw_text(0x190, y, color, time_str);
    }

    void draw_text_with_time(u32 x, u32 y, u32 color, const char *str, const int64_t time) {
        draw_text_with_time(x, y, color, str, time / 1000000.0f);
    }

    void intro() {
        appletLockExit();

        PlFontData font;

        plGetSharedFontByType(&font, PlSharedFontType_Standard);

        FT_Init_FreeType(&library);
        FT_New_Memory_Face(library, static_cast<FT_Byte *>(font.address), font.size, 0, &face);
        FT_Set_Char_Size(face, 0, 6*64, 300, 300);

        framebufferCreate(&fb, nwindowGetDefault(), FB_WIDTH, FB_HEIGHT, PIXEL_FORMAT_RGBA_8888, 2);
        framebufferMakeLinear(&fb);
        framebuf = (u32 *)framebufferBegin(&fb, &stride);
        framebuf_width = stride / sizeof(u32);
        memset(framebuf, 0, stride*FB_HEIGHT);
        framebufferEnd(&fb);

        draw_text(0x010, 0x020, YELLOW, "Lockpick! by shchmue");
        draw_text(0x190, 0x020, YELLOW, "Note:");
        draw_text(0x1e0, 0x020, YELLOW, "Lockpick can only dump keys 00-05 (or 00-06 on 6.2.0)");
        draw_text(0x1e0, 0x040, CYAN, "Lockpick_RCM");
        draw_text(0x2a0, 0x040, YELLOW, "can get newer keys on firmware 7.0.0+!");

        draw_set_rect(814, 452 + 42 * 0, 450, 42, FLAG_RED);
        draw_set_rect(814, 452 + 42 * 1, 450, 42, FLAG_ORANGE);
        draw_set_rect(814, 452 + 42 * 2, 450, 42, FLAG_YELLOW);
        draw_set_rect(814, 452 + 42 * 3, 450, 42, FLAG_GREEN);
        draw_set_rect(814, 452 + 42 * 4, 450, 42, FLAG_BLUE);
        draw_set_rect(814, 452 + 42 * 5, 450, 42, FLAG_VIOLET);

        if (  !(envIsSyscallHinted(0x60) &&     // svcDebugActiveProcess
                envIsSyscallHinted(0x63) &&     // svcGetDebugEvent
                envIsSyscallHinted(0x65) &&     // svcGetProcessList
                envIsSyscallHinted(0x69) &&     // svcQueryDebugProcessMemory
                envIsSyscallHinted(0x6a))) {    // svcReadDebugProcessMemory
            draw_text(0x190, 0x20, RED, "Error: Please run with debug svc permissions!");
            wait_to_exit();
        }

        draw_text(0x10, 0x060, CYAN, "Get Tegra keys...");
        draw_text(0x10, 0x080, CYAN, "Get keys from memory...");
        draw_text(0x10, 0x0a0, CYAN, "Get master keys...");
        draw_text(0x10, 0x0c0, CYAN, "Derive remaining keys...");
        draw_text(0x10, 0x0e0, CYAN, "Saving keys to keyfile...");
        draw_text(0x10, 0x110, CYAN, "Total time elapsed:");

        update_display();
    }

    void get_tegra_keys(Key &sbk, Key &tsec, Key &tsec_root) {
        // support Hekate dump
        if (std::filesystem::exists("/backup")) {
            for (auto &p : std::filesystem::recursive_directory_iterator("/backup")) {
                if (p.is_regular_file()) {
                    if (!sbk.found() && (p.file_size() == 0x2fc || p.file_size() == 0x300) &&
                        ((p.path().filename().string().substr(0, 5).compare("fuses") == 0) ||
                        (p.path().filename().string().substr(0, 11).compare("fuse_cached") == 0)))
                    {
                        FILE *fuse_file = fopen(p.path().c_str(), "rb");
                        if (!fuse_file) continue;
                        byte_vector temp_key(0x10);
                        fseek(fuse_file, 0xa4, SEEK_SET);
                        fread(temp_key.data(), 0x10, 1, fuse_file);
                        sbk = Key("secure_boot_key", 0x10, temp_key);
                        fclose(fuse_file);
                    }
                    else if (!tsec.found() && (p.file_size() == 0x20 || p.file_size() == 0x30) &&
                        (p.path().filename().string().substr(0, 4).compare("tsec") == 0))
                    {
                        FILE *tsec_file = fopen(p.path().c_str(), "rb");
                        if (!tsec_file) continue;
                        byte_vector temp_key(0x10);
                        fread(temp_key.data(), 0x10, 1, tsec_file);
                        tsec = Key("tsec_key", 0x10, temp_key);
                        fread(temp_key.data(), 0x10, 1, tsec_file);
                        tsec_root.find_key(temp_key);
                        fclose(tsec_file);
                    }
                }
                if (sbk.found() && tsec.found())
                    return;
            }
        }
        // support biskeydump v7 dump
        if (std::filesystem::exists("/device.keys")) {
            FILE *key_file = fopen("/device.keys", "r");
            char line[0x100];
            while (fgets(line, sizeof(line), key_file) && !(sbk.found() && tsec.found())) {
                if (strncmp("secure_boot_key", line, 15) == 0)
                    sbk = Key("secure_boot_key", 0x10, key_string_to_byte_vector(line));
                else if (strncmp("tsec_key", line, 8) == 0)
                    tsec = Key("tsec_key", 0x10, key_string_to_byte_vector(line));
            }
            fclose(key_file);
        }
    }

    void wait_to_exit() {
        draw_text(0x10b, 0x24b, YELLOW, ">> Press + to exit <<");

        while(appletMainLoop()) {
            hidScanInput();
            u64 kDown = hidKeysDown(CONTROLLER_P1_AUTO);
            if (kDown & KEY_PLUS) break;

            update_display();
        }

        framebufferClose(&fb);
        FT_Done_Face(face);
        FT_Done_FreeType(library);

        appletUnlockExit();
    }

    void update_display() {
        framebufferBegin(&fb, &stride);
        framebufferEnd(&fb);
    }

    byte_vector key_string_to_byte_vector(std::string key_string) {
        key_string = key_string.substr(key_string.find('=') + 2);
        byte_vector temp_key((key_string.size() - 1) / 2);
        for (size_t i = 0; i < temp_key.size() - 1; i += 8)
            *reinterpret_cast<u64 *>(temp_key.data() + i) = __bswap64(strtoul(key_string.substr(i * 2, 0x10).c_str(), NULL, 16));
        return temp_key;
    }
}